### DISCUSSION TITLE
Maximum allowed cache size for Dynamo DB with Laravel Vapor

### DISCUSSION THREAD

**Discussion by Kevmue** [spatie/laravel-permission#1773] [Q&A]
When deploying one of our projects to Laravel Vapor we encountered an issue with the laravel-permission extension:

Error: 
`Client error: 'POST https://dynamodb.eu-central-1.amazonaws.com' resulted in a '400 Bad Request' response: {"__type":"com.amazon.coral.validate#ValidationException","message":"Item size has exceeded the maximum allowed size"} `

I've tracked down the issue to the getPermissions method in the PermissionsRegistrar.php. The caching of the permissions seems to exceed the allowed 400KB by DynamoDB.

Is there a way to resolve this except reducing the number of permissions or switching to redis?

**Comment by drbyte** [comment]
I have no direct experience to assist with this exact environment.

Is the cached permission size "expected"? (thinking here: do you have thousands of permissions, and intertwined relations amongst them with users and roles and other models; understanding the numbers of these might help understand scope)
Are other caching segments of your application also causing cache payloads too large for your setup as well?

**Comment by drbyte** [comment]
Note: You can set the config for `permission.cache.store` to any of the config values in `cache.stores` that you've defined in your Laravel application. Perhaps this may give you an option for relocating the cache to something that supports larger sizes?

**Comment by Kevmue** [comment]
Thanks for the response.

I don't think that we have an extraordinary  number of permissions / roles.
There are actually 5 roles and 84 permissions with a max of 3 roles assigned to them each.

Here a short snippet I used to determine the size 

```php

$permissions = Spatie\Permission\Models\Permission::getPermissions()
$strlen      =  strlen(serialize($permissions)); // -> 514280 -> ~514kB I guess
$count       =  count($permissions); // -> 84
```

Thanks for your suggestion for switching the cache store. I already thought about that. However, the only alternative is to use a redis cluster. This will add ~50$ of additional monthly cost to that setup which is rather a lot for the size of this and probably many other projects. 

Is there a way to deactivate caching or any other idea how to resolve this?

**Comment by drbyte** [comment]
> Is there a way to deactivate caching or any other idea how to resolve this?

You could set `permission.cache.store` to `'array'` so that it only "caches" to RAM during the current request, no persisting to any database.
Then test to see whether it changes performance significantly at all.

**Comment by jonJKF** [comment]
I'm running into the same issue here, disabling caching by updating my config/permissions.php and adding ```'store' => 'array',``` to the caching section seems to allow me to bypass this error but is surely not ideal for performance.

For reference my application has 25 roles and about over 600 permissions (50+ database models each with their own view,  create, edit, delete permissions). It sounds like a lot but really it isn't, it just an average web app that has a slightly above average relational schema.

There must be a way to be able to cache and persist roles and permissions to improve performance whilst avoiding creating a single 400KB+ object?

**Comment by iAmKevinMcKee** [comment]
I'm getting this issue as well. Would love to know if there are any solutions aside from setting the cache store to array

**Comment by michaelnguyen2021** [comment]
@Kevmue don't set up a VPC to use AWS redis because of Vapor. Sign up an account with RedisLabs. It may be cheaper than using DynamoDB. I don't have a lot of permission with my Vapor. I m totally not happy with DynamoDB monthly fee. It is roughly $15. I don't use cache anywhere besides this package permission cache.

**Comment by drbyte** [comment]
NOTE: v`4.3.0` now produces a much smaller cache

**Comment by gustavo-ticto** [comment]
I'm facing the same issue here. I'll prob just plug in a redis store for that matter..

**Reply by erikn69** [reply]
Which version are you using?

**Reply by gustavo-ticto** [reply]
^5.5

**Reply by parallels999** [reply]
How big is your permisions cache size?

**Reply by gustavo-ticto** [reply]
Just found out, it was my fault. I've been using roles for impersonate users on the app. So after pruning those roles the cache on Dynamodb got back to work.

But if my app grows to a lot of roles, the cache won't fit

**Reply by gustavo-ticto** [reply]
> How big is your permisions cache size?

we have around 200 permissions. But the user can create roles and attach permissions to them.

**Reply by parallels999** [reply]
> we have around 200 permissions.

I mean on kb sizes
Also you can make a PR for smaller cache, example: https://github.com/spatie/laravel-permission/pull/1799, https://github.com/spatie/laravel-permission/pull/1912/commits/2ae7b23c210de385d728d9954e74894b980e4a80
